contents/java/JavaConcurrency.md

```diff
@@ -0,0 +1,149 @@
+    * [Operations on Threads](#operations-on-threads)
+    * [Thread Interference](#thread-interference)
+    * [Executors](#executors)
+* [Disadvantages of Java Concurrency](#disadvantages-of-java-concurrency)
+* [Resources](#resources)
+    * [Further Reading](#further-reading)
+
+## What is Java Concurrency
+_Concurrency_ is the ability to run several programs or several parts of a program **out-of-order**, in an **interleaved fashion**. Simply put, if a program is running concurrently, the program execution will execute one part of the program, pause it, execute another part and repeat. This is different from _Parallelism_ in which parts of the program are executed at the **same time**.
+
+Java Concurrency enables you to use perform tasks using multi-threading in your code and therefore:
+1. improving performance
+1. improving responsiveness
+1. allowing you to take advantage of multi-core processors
+
+## Getting Started
+There are many tutorials that cover Java Concurrency in-depth, such as the [Java tutorial by Oracle](https://docs.oracle.com/javase/tutorial/essential/concurrency/). Instead, this chapter will provide an overview and things to take note of.
+
+### Understanding Threads
+First off, a process is simply a program in execution. It contains at least one thread and has it's own memory space.
+
+A thread: 
+* is a lightweight process that exists within a process 
+* is an independent path of execution through program code 
+* has its own call stack and can access shared data of other threads in the same process. 
+
+When multiple threads execute code in parallel of other threads running, it gives rise to multi-threading.
+
+### Creating and Starting Threads
+There are two ways to create a thread in Java:
+ 
+1. Extend the `Thread` class
+1. Implement the `Runnable` interface
+
+After which, override the `run()` method which contains the code that will be executed when the thread starts. 
+
+Depending on how you created the `Thread` you can either create the class that extends the `Thread` class or pass the class that implements the `Runnable` interface into the Thread constructor and then start it.
+
+When deciding which method to use to create a `Thread`, it is always advisable to implement the `Runnable` interface as this results in composition which will allow your code to be more loosely coupled as compared to inheritance. Furthermore, you can extend another class if need be.
+
+### Operations on Threads
+After creating and starting threads, you can carry out operations on them. There are several such operations you can use to manipulate threads:
+
+* Making a thread sleep for a given amount of time. This is useful for pausing the execution of a certain thread for a certain amount of time to free up resources for other threads. The sleeping thread can be interrupted by another thread, which can be a way of communicating that the other thread is done with it's work allowing the current thread to start.  
```

Comment:
Just wondering, what's the difference between using a lock for the threads vs handling `InterruptedException`? It seems that handling `InterruptedException` isn't a good way of "communicating that the other thread is done with it's work allowing the current thread to start" because exceptions ought to be exceptional; it shouldn't be part of control flow.

Reply:
Ahh, my bad :sweat_smile: . A lock should be a better way of `communicating that the other thread is done with it's work allowing the current thread to start`. `Thread.sleep` is instead **an efficient means of making processor time available to the other threads of an application or other applications that might be running on a computer system**.